crates/telio-starcast/src/transport.rs

```diff
@@ -0,0 +1,231 @@
+        socket_pool: Arc<SocketPool>,
+        packet_chan: Chan<Vec<u8>>,
+        multicast_ips: Vec<IpNetwork>,
+    ) -> Result<Self, Error> {
+        let transport_socket = socket_pool
+            .new_internal_udp(SocketAddr::new(meshnet_ip, MULTICAST_TRANSPORT_PORT), None)
+            .await
+            .map_err(|e| {
+                telio_log_error!("Failed to bind multicast socket: {:?}", e);
+                Error::SocketBindError(e.to_string())
+            })?;
+        let transport_socket = Arc::new(transport_socket);
+
+        Ok(Self {
+            task: Task::start(State {
+                transport_socket,
+                packet_chan,
+                multicast_ips,
+                recv_buffer: vec![0; MAX_PACKET_SIZE],
+                nat: StarcastNat::new(vpeer_ip.ipv4, vpeer_ip.ipv6),
+                peers: Vec::new(),
+            }),
+        })
+    }
+
+    /// Update runtime values of the transport component
+    pub async fn configure(&self, config: Config) -> Result<(), Error> {
+        task_exec!(&self.task, async move |state| {
+            state.configure(config);
+            Ok(())
+        })
+        .await?;
+        Ok(())
+    }
+
+    /// Stop the transport component
+    pub async fn stop(self) {
+        let _ = self.task.stop().await.resume_unwind();
+    }
+}
+
+struct State {
+    transport_socket: Arc<UdpSocket>,
+    packet_chan: Chan<Vec<u8>>,
+    recv_buffer: Vec<u8>,
+    nat: StarcastNat,
+    peers: Vec<Peer>,
+    multicast_ips: Vec<IpNetwork>,
+}
+
+impl State {
+    fn configure(&mut self, config: Config) {
+        self.peers = config.peers;
+    }
+
+    async fn handle_local_multicast_packet(&mut self, packet: Vec<u8>) -> Result<(), Error> {
+        let failed_peers = join_all(self.peers.iter().map(|peer| {
+            self.transport_socket
+                .send_to(
+                    &packet,
+                    SocketAddr::new(peer.addr, MULTICAST_TRANSPORT_PORT),
+                )
+                .map_err(|_| peer.public_key)
+        }))
+        .await
+        .into_iter()
+        .filter_map(|res| match res {
+            Ok(_) => None,
+            Err(pk) => Some(pk),
+        })
+        .collect::<Vec<_>>();
+
+        if failed_peers.is_empty() {
+            Ok(())
+        } else {
+            Err(Error::FanoutFailed(failed_peers))
+        }
+    }
+
+    async fn handle_mapped_unicast_packet(&mut self, mut packet: Vec<u8>) -> Result<(), Error> {
+        self.nat
+            .translate_outgoing(&mut packet)
+            .map_err(Error::NatError)?;
+        self.transport_socket
+            .send(&packet)
+            .await
+            .map(|_| ())
+            .map_err(|_| Error::SocketSendError)
+    }
+
+    async fn handle_remote_packet(&mut self, mut packet: Vec<u8>) -> Result<(), Error> {
+        self.nat
+            .translate_incoming(&mut packet)
+            .map_err(Error::NatError)?;
+        self.packet_chan
+            .tx
```

Comment:
This should be via permit, recieved on select arm, as we are sending packets back and forth, we would run into constant error loop

Reply:
Ok you are not using try_send, you are using a send. This has a change to casue a deadlock then to task are both sending and reading from each other, if not done via permits.

These test's is an artificial example how this can happen: https://github.com/NordSecurity/libtelio/blob/main/crates/telio-task/src/io.rs#L126